File: C#/Menu.cs
using System;
using System.Collections.Generic;
using System.Text;
using DSharpPlus.Entities;


namespace CSDNDISCORD
{

    class Menu
    {
        public string name { get; private set; }
        public string description { get; private set; }
        public string emoji { get; private set; }
        public List<Menu> childMenus { get; private set; }
        public Func<Player, bool> operation { get; private set; }
        public Func<Player, bool> counterOperation { get; private set; }
        public bool showDescription { get; private set; }
        public Menu parentMenu { get; private set; } = null;
        public bool empty { get; private set; }

        public static int spaces = 8;
        public static string backemoji = ":back:";

        public Menu(string title, string description, string emoji, List<Menu> childMenus, Func<Player, bool> action = null, Func<Player, bool> counterAction = null, bool empty=false,
            bool showDescription = true)
        {
            this.name = title;
            this.description = description;
            this.emoji = emoji;
            this.childMenus = childMenus;
            this.operation = action;
            this.counterOperation = counterAction;
            this.showDescription = showDescription;
            this.empty = empty;

            foreach(Menu m in this.childMenus)
            {
                m.parentMenu = this;
            }

        }


        public async void Send(DiscordChannel channel, Player sender)
        {
            bool sendMessage = true;
            if (this.operation != null) sendMessage = this.operation.Invoke(sender);

            if (!sendMessage || empty) return;


            string title = this.name;
            if (this.showDescription == true)
            {

                title += " - ";
                title += this.description;
            }

            string description = "";
            foreach (Menu m in this.childMenus)
            {
                description += m.emoji;
                description += new String('d', spaces).Replace("d", " - ");
                description += "`" + m.name + "`";
                description += "\n";
            }

            DiscordMessage msg = await channel.SendMessageAsync(title + "\n" + description);

            List<string> reactions = new List<string>();

            foreach (Menu m in this.childMenus)
            {
                await msg.CreateReactionAsync(DiscordEmoji.FromName(Program.bot.Client, m.emoji));
                reactions.Add(m.emoji);
            }
            if (this.parentMenu != null)
            {
                await msg.CreateReactionAsync(DiscordEmoji.FromName(Program.bot.Client, backemoji));
                reactions.Add(backemoji);
            }

            

            var reactionResult = await Program.bot.Interactivity.WaitForReactionAsync(r => !r.User.IsBot && (reactions.Contains(r.Emoji.GetDiscordName())));

            if (reactionResult.Result != null)
            {
                if (reactionResult.Result.Emoji.GetDiscordName() == backemoji)
                {
                    this.counterOperation.Invoke(sender);
                    this.parentMenu.Send(channel, sender);
                }
                else
                {
                    foreach (Menu m in this.childMenus)
                    {
                        if (m.emoji == reactionResult.Result.Emoji.GetDiscordName())
                        {
                            m.Send(channel, sender);
                            break;
                        }
                    }
                }
            }
            await msg.DeleteAsync();
        }

        public void AddChild(Menu child)
        {
            this.childMenus.Add(child);
        }
    }

}

File: PYTHON/Menu.py
import discord
import GlobalVars
from typing import Callable, List, Awaitable
from Player import Player

backEmoji = "🔙"


class Menu:

    def __init__(self, title:str, emoji, childMenus, description=None,
                 action : Callable[[discord.TextChannel, Player], Awaitable[bool]]=None,
                 counterAction : Callable[[discord.TextChannel, Player], Awaitable[bool]]=None, **properties):

        self.title : str = title
        self.description : str = description if description != None and description else ""
        self.emoji : str = emoji
        self.action : Callable[[discord.TextChannel, Player], Awaitable[bool]] = action
        self.counterAction : Callable[[discord.TextChannel, Player], Awaitable[bool]] = counterAction

        self.embed : discord.Embed = properties.get('embed', None)
        self.deleteAfter : float = properties.get('deleteAfter', None)
        self.childMenus : List[Menu] = childMenus
        self.parentMenu : Menu = None
        self.showDescription = properties.get('showDescription', False)

        for m in self.childMenus:
            m.parentMenu = self

    async def Send(self, channel : discord.TextChannel, sender : Player):

        if self.action != None:
            result = await self.action(channel, sender)
            if not result:
                return

        spaces = 8

        title = f"`{self.title}{(' - ' + self.description if self.description != '' and self.showDescription else '')}`"
        description = "\n".join([f"{c.emoji + spaces * ' - '}`{c.title}`"
                                         for c in self.childMenus])
        sendMessage = title + "\n" + description

        message = await channel.send(content=sendMessage,
                                     embed=self.embed,
                     delete_after=self.deleteAfter)

        for m in self.childMenus:
            await message.add_reaction(m.emoji)
        if self.parentMenu != None:
            await message.add_reaction(backEmoji)


        menuEmojis = [m.emoji for m in self.childMenus]
        reaction, user = await GlobalVars.bot.wait_for('reaction_add', check=lambda r, u : ((r.emoji in menuEmojis) or (r.emoji == backEmoji)) and u == sender.author)

        await message.delete(delay=0.35)

        if reaction.emoji == backEmoji:
            if self.counterAction != None:
                sendParent = await self.counterAction(sender)
                if sendParent:
                    await self.parentMenu.Send(channel, sender)
                return

            await self.parentMenu.Send(channel, sender)
        else:
            await [m for m in self.childMenus if m.emoji == reaction.emoji][0].Send(channel, sender)

